### ISSUE TITLE
Assertion failure in `debug_heap.cpp` when using `xlnt::cell_reference::to_string()`

### ISSUE THREAD

**Issue by nicosafull** [xlnt-community/xlnt#13]
I'm experiencing an assertion failure in `debug_heap.cpp` when using the `xlnt::cell_reference::to_string()` method in the following sample code:

```cpp
#include <xlnt/xlnt.hpp>

int main()
{
    for (_ulong xlntCol = 0; xlntCol < 1; ++xlntCol)
    {
        std::string cellReferenceValue = xlnt::cell_reference(xlntCol + 1, 2).to_string();
    }
    return 0;
}
```

The assertion failure occurs on the line `_ASSERTE(__acrt_first_block == header);` in `debug_heap.cpp`. This suggests that there is a problem with the memory management in the CRT (C Runtime) library.

**Steps to Reproduce:**

1. Compile the `xlnt` library with the following settings:
	* Target platform: x86
	* C++ language standard: C++14
	* Runtime library: Multi-threaded Debug (/MTd)
2. Compile the sample code with the same settings as the `xlnt` library.
3. Run the sample code.

**Expected Behavior:**

The sample code should run without any assertion failures.

**Actual Behavior:**

The sample code fails with an assertion failure on the line `_ASSERTE(__acrt_first_block == header);` in `debug_heap.cpp`.

**Environment:**

* Operating system: Windows 11
* Compiler: Visual Studio 2022
* `xlnt` library version: latest as of today, compiled and generated as a shared library by me through CMAKE

**Additional Information:**

I've confirmed that the issue is not caused by a memory corruption in my code, as I've been able to reproduce the issue with the sample code provided. I've also tried using a different version of the `xlnt` library, but the issue persists.

I've also tried reducing the code further to isolate the issue, but the issue still occurs when calling `xlnt::cell_reference::to_string()`.

settings in CMAKE:
![image](https://github.com/user-attachments/assets/5017b233-160e-4aef-a4e1-6258dbe551d6)

**Comment by nicosafull** [comment]
seems this has to do with it as well https://github.com/google/googletest/issues/2299#issuecomment-629664435

Anyways, compiling it as static works fine :)

**Comment by m7913d** [comment]
It seems to me that you already solved your problem. Personally, I'm not very familiar with the `/MT[d]` option. As far as I know:
 - `/MD[d]` (=dynamic runtime library) is the safest option
 - `/MT[d]` was only enabled/tested in Wt in combination with static libraries. Currently, we aren't using it anymore in CI / unit testing.
Why did you explicitly enabled this option without static building?

**Comment by doomlaur** [comment]
Just to be sure, @nicosafull: are you 100% sure that they are exactly the same settings? For me, it sounds like you are running into this issue: [Is it possible to mix static, multi-threaded, DLL libraries in one project?](https://stackoverflow.com/questions/2285408/is-it-possible-to-mix-static-multi-threaded-dll-libraries-in-one-project)

If that is not the issue, let me know and I'll try it with Visual Studio 2022 17.11 on Windows 10 and see if I can reproduce the issue.

**Comment by nicosafull** [comment]
> Just to be sure, @nicosafull: are you 100% sure that they are exactly the same settings? For me, it sounds like you are running into this issue: [Is it possible to mix static, multi-threaded, DLL libraries in one project?](https://stackoverflow.com/questions/2285408/is-it-possible-to-mix-static-multi-threaded-dll-libraries-in-one-project)
> 
> If that is not the issue, let me know and I'll try it with Visual Studio 2022 17.11 on Windows 10 and see if I can reproduce the issue.

yes, I made sure to complie and generate the build with same options (MT[d]) as well as my sample project, you are correct that if they are mixed it won't work, in fact I can't even invoke workbook.load if they don't match. I don't dismiss any special setting in my project as it's super old and heavily coupled with other things.

But as I've said, compiling it as static did it for me and I don't get such warning:).

If you cannot replicate I think it's fine to close the issue, may help someone else with an actual problem

**Comment by doomlaur** [comment]
I just tried it out with Visual Studio 2022 17.11.1 on Windows 10. I simply cloned our current repository, in CMake I selected `STATIC_CRT` (so that `Multi-threaded Debug (/MTd)` is automatically selected in Visual Studio). I then created a new project, added XLNT as a dependency, and set the project to `/MTd` as well. The result: XLNT will be built as a DLL with statically linked CRT, while my test project will load the XLNT DLL and will also link the CRT statically itself.

I can fully reproduce your issue. I have also tried out the Address Sanitizer and it showed me the issue precisely. But first things first:

1. I used the following code:
```cpp
#include <string>
#include <xlnt/xlnt.hpp>

int main()
{
    for (xlnt::column_t xlntCol = 0u; xlntCol < 1u; ++xlntCol)
    {
        std::string cellReferenceValue = xlnt::cell_reference(xlntCol + 1, 2).to_string();
    }
    return 0;
}
```

2. The call to `xlnt::cell_reference::to_string` creates a string (obviously). The string returned by XLNT in my case is simply `A2`, containing 2 characters, meaning that it's way below the limit of 15 characters for which it uses Short String Optimization (SSO). In other words, the memory is allocated on the stack and has an internal size of 16 bytes (due to \0 at the end). However, the C++ Standard Library still does allocate 16 bytes on the heap for a `_Container_proxy`:
![msvc_new_string_xlnt](https://github.com/user-attachments/assets/f73c433e-a77a-4816-a685-88f2c14831fc)

3. After XLNT has allocated memory for the string and after creating the actual string, it is returned to our program. However, at the end of the loop, the string it is deallocated and ASAN complains about it: 
![msvc_delete_string_xlnt](https://github.com/user-attachments/assets/e619cfc3-2186-4e74-bfa6-edde36ad4c83)

So apparently the issue is that memory is allocated in an **_external DLL_** (by XLNT) that your project (and my test project) cannot delete correctly. However, this seems to be an issue with how MSVC works, as the whole memory allocation of `std::string` is not controlled by XLNT.

Personally, I do not see how we can solve this issue. The only thing I can image is to block XLNT from building the CMake project while it is compiled as a DLL but with static CRT. However, I'm open for suggestions, so if you have a better idea, please let me know!

**Comment by m7913d** [comment]
Personally, I would adapt cmake to ignore the STATIC_CRT option if  it is not a static build or generate an error message in that case.

**Comment by doomlaur** [comment]
Honestly, I did not expect us to have this issue. I knew about issues when having different CRTs (e.g. we had some issues where Python used another CRT and the FILE * was not compatible), however, I did not know that it's also an issue even if the CRT is the very same (both you and me used the same version for both XLNT and the project). I'm always simply linking the CRT dynamically, [as recommended by Microsoft](https://learn.microsoft.com/en-us/cpp/windows/choosing-a-deployment-method?view=msvc-170).

Just to be clear: it's not an issue if XLNT or a different project allocate memory themselves and use it internally, but it is a problem if they move the memory across DLL boundaries.

Although I don't think that supporting this combination of configuration options is a good idea for XLNT's future (read below why), just out of curiosity, I quickly looked at what changes would be necessary to avoid memory issues. In other words, XLNT would NEVER be allowed to return anything that allocated memory, be it from C++ (using std::string, std::vector or anything else), or C (C arrays or functions of the C Standard Library returning a newly allocated array). From a quick search, we would need to replace the following in all XLNT headers where a copy of the data is returned (not a reference to internal data):
- std::string
- std::wstring
- std::vector
- std::unordered_map

Not only would this be a huge amount of work, but we would also probably create containers that have a lower quality overall than the ones provided by C++. Believe me, I know custom string classes from other projects, and people either create them for a very specific use case and never use them outside these (then it's fine), or they think it's not much work and end up effectively rewriting the whole std::string after using the custom string class long enough (because there is a huge number of operations that people do on strings). And this is just for strings, not mentioning `std::vector` and `std::unordered_map`. Maybe it would be possible to use a custom `std::allocator` for these containers, but even them, we would neither be allowed to use `new` from C++ nor `malloc` from C, but forced to allocate memory using OS-specific functions, e.g. [`HeapAlloc`](https://learn.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc) on Windows.

While I couldn't find that much information about our issue, but I could find the following:

1. [Microsoft - Potential errors passing CRT objects across DLL boundaries](https://learn.microsoft.com/en-us/cpp/c-runtime-library/potential-errors-passing-crt-objects-across-dll-boundaries?view=msvc-170)
2. [Swift forums - Proper Support for MSVCRT linkage](https://forums.swift.org/t/proper-support-for-msvcrt-linkage/24423)

Both links don't recommend doing what we are currently testing here.

Last but not least, I also want to say it from another perspective: parsing files of the Office Open XML specification is VERY hard to do correctly. This library does an amazing work (thanks @tfussel), but has its issues (which we are currently working on, step by step). The whole specification has multiple parts, and some of the parts have over 5000 pages. In other words, I think that we already have a lot of work to do, and working on supporting this use case would mean a lot of work for little benefit for most XLNT users.

As always, I'm open for discussion, but at this moment, I see a very small advantage but a whole lot of disadvantages which would drastically slow down the XLNT development unnecessarily. Let me know if you think otherwise :slightly_smiling_face:

**Comment by m7913d** [comment]
@nicosafull Are you interested in creating a PR to solve this issue?

**Comment by doomlaur** [comment]
@nicosafull Just for the record, this issue has been automatically closed after merging #17. Feel free to still express your opinion if you have a different opinion than ours, and feel free to reopen the issue in case you want to send a Pull Request :wink:
